framed-messages/src/lib.rs

```diff
@@ -0,0 +1,97 @@
+pub mod framed;
+
+use bytes::BytesMut;
+use futures_util::{SinkExt, TryStreamExt};
+use iroh::{
+    endpoint::Connection,
+    protocol::{AcceptError, ProtocolHandler},
+};
+use serde::{Deserialize, Serialize};
+
+use crate::framed::FramedBiStream;
+
+/// Each protocol is identified by its ALPN string.
+///
+/// The ALPN, or application-layer protocol negotiation, is exchanged in the connection handshake,
+/// and the connection is aborted unless both nodes pass the same bytestring.
+pub const ALPN: &[u8] = b"iroh/examples/messages/0";
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub struct Move {
+    pub from: (u8, u8),
+    pub to: (u8, u8),
+}
+
+impl Move {
+    pub async fn send(&self, stream: &mut FramedBiStream) -> anyhow::Result<()> {
+        let encoded_move = postcard::to_extend(self, BytesMut::new())?.freeze();
+        stream.write.send(encoded_move).await?;
+        Ok(())
+    }
+
+    pub async fn recv(stream: &mut FramedBiStream) -> anyhow::Result<Move> {
+        let Some(encoded_move) = stream.read.try_next().await? else {
+            anyhow::bail!("unexpected end of stream");
+        };
+        let mv = postcard::from_bytes(&encoded_move)?;
+        Ok(mv)
+    }
+}
+
+async fn make_moves(stream: &mut FramedBiStream) -> anyhow::Result<()> {
+    let mv = Move::recv(stream).await?;
+    println!("got move: {mv:?}");
+
+    // Respond with some move
+    let mv = Move {
+        from: (5, 7),
+        to: (5, 6),
+    };
+    mv.send(stream).await?;
+
+    let mv = Move::recv(stream).await?;
+    println!("got move: {mv:?}");
+
+    // And respond with another one
+    let mv = Move {
+        from: (5, 8),
+        to: (5, 7),
+    };
+    mv.send(stream).await?;
+
+    Ok(())
+}
```

Comment:
seems this is defined twice? once here & once in `main.rs`

Reply:
They're technically different moves, and one first receives, the other first sends a move.

Maybe I'll rename them `make_moves_white` and `make_moves_black` or sth